man/chpst.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "chpst" "8" 
.SH NAME
chpst \- runs a program with a changed process state
.SH SYNOPSIS
\f[B]chpst\f[R] [\-vVPFI012N] [\-u \f[I]user\f[R]] [\-U \f[I]user\f[R]]
[\-b \f[I]argv0\f[R]] [\-e \f[I]dir\f[R]] [\-/ \f[I]root\f[R]] [\-C
\f[I]pwd\f[R]] [\-n \f[I]inc\f[R]] [\-l|\-L \f[I]lock\f[R]] [\-A
\f[I]seconds\f[R]] [\-m \f[I]bytes\f[R]] [\-d \f[I]bytes\f[R]] [\-o
\f[I]n\f[R]] [\-p \f[I]n\f[R]] [\-f \f[I]bytes\f[R]] [\-c
\f[I]bytes\f[R]] [\-t \f[I]seconds\f[R]] \f[I]prog\f[R]
.SH DESCRIPTION
\f[I]prog\f[R] consists of one or more arguments.
.PP
\f[B]chpst\f[R] changes the process state according to the given
options, and runs \f[I]prog\f[R].
.SH OPTIONS
.TP
\f[B]\-u\f[R] \f[I][:]user[:group]\f[R]
setuidgid.
Set uid and gid to the \f[I]user\f[R]\[aq]s uid and gid, as found in
\f[I]/etc/passwd\f[R].
If \f[I]user\f[R] is followed by a colon and a \f[I]group\f[R], set the
gid to \f[I]group\f[R]\[aq]s gid, as found in \f[I]/etc/group\f[R],
instead of \f[I]user\f[R]\[aq]s gid.
If \f[I]group\f[R] consists of a colon\-separated list of group names,
\f[B]chpst\f[R] sets the group ids of all listed groups.
If \f[I]user\f[R] is prefixed with a colon, the \f[I]user\f[R] and all
\f[I]group\f[R] arguments are interpreted as uid and gids respectively,
and not looked up in the password or group file.
All initial supplementary groups are removed.
.TP
\f[B]\-U\f[R] \f[I][:]user[:group]\f[R]
envuidgid.
Set the environment variables $UID and $GID to the \f[I]user\f[R]\[aq]s
uid and gid, as found in \f[I]/etc/passwd\f[R].
If \f[I]user\f[R] is followed by a colon and a \f[I]group\f[R], set $GID
to the \f[I]group\f[R]\[aq]s gid, as found in \f[I]/etc/group\f[R],
instead of \f[I]user\f[R]\[aq]s gid.
If \f[I]user\f[R] is prefixed with a colon, the \f[I]user\f[R] and
\f[I]group\f[R] arguments are interpreted as uid and gid respectively,
and not looked up in the password or group file.
.TP
\f[B]\-b\f[R] \f[I]argv0\f[R]
argv0.
Run \f[I]prog\f[R] with \f[I]argv0\f[R] as the 0th argument.
.TP
\f[B]\-e\f[R] \f[I]dir\f[R]
envdir.
Set various environment variables as specified by files in the directory
\f[I]dir\f[R]: If \f[I]dir\f[R] contains a file named \f[I]k\f[R] whose
first line is \f[I]v\f[R], \f[B]chpst\f[R] removes the environment
variable \f[I]k\f[R] if it exists, and then adds the environment
variable \f[I]k\f[R] with the value \f[I]v\f[R].
The name \f[I]k\f[R] must not contain =.
Spaces and tabs at the end of \f[I]v\f[R] are removed, and nulls in
\f[I]v\f[R] are changed to newlines.
If the file \f[I]k\f[R] is empty (0 bytes long), \f[B]chpst\f[R] removes
the environment variable \f[I]k\f[R] if it exists, without adding a new
variable.
.TP
\f[B]\-/\f[R] \f[I]root\f[R]
chroot.
Change the root directory to \f[I]root\f[R] before starting
\f[I]prog\f[R].
.TP
\f[B]\-C\f[R] \f[I]pwd\f[R]
chdir.
Change the working directory to \f[I]pwd\f[R] before starting
\f[I]prog\f[R].
When combined with \-/, the working directory is changed after the
chroot.
.TP
\f[B]\-n\f[R] \f[I]inc\f[R]
nice.
Add \f[I]inc\f[R] to the \f[B]nice\f[R](2) value before starting
\f[I]prog\f[R].
\f[I]inc\f[R] must be an integer, and may start with a minus or plus.
.TP
\f[B]\-l\f[R] \f[I]lock\f[R]
lock.
Open the file \f[I]lock\f[R] for writing, and obtain an exclusive lock
on it.
\f[I]lock\f[R] will be created if it does not exist.
If \f[I]lock\f[R] is locked by another process, wait until a new lock
can be obtained.
.TP
\f[B]\-L\f[R] \f[I]lock\f[R]
The same as \-l, but fail immediately if \f[I]lock\f[R] is locked by
another process.
.TP
\f[B]\-A\f[R] \f[I]seconds\f[R]
alarm signal.
Run \f[I]prog\f[R] under an alarm, delivering a SIGALRM after
\f[I]seconds\f[R] seconds.
If \f[I]seconds\f[R] is zero, clear any inherited alarm.
When combined with \f[B]\-F\f[R], the alarm is set before entering the
new namespace, so SIGALRM is relayed through \f[B]chpst\f[R] to all
processes in \f[I]prog\f[R]\[aq]s namespace, including its children.
.TP
\f[B]\-m\f[R] \f[I]bytes\f[R]
limit memory.
Limit the data segment, stack segment, locked physical pages, and total
of all segment per process to \f[I]bytes\f[R] bytes each.
.TP
\f[B]\-d\f[R] \f[I]bytes\f[R]
limit data segment.
Limit the data segment per process to \f[I]bytes\f[R] bytes.
.TP
\f[B]\-o\f[R] \f[I]n\f[R]
limit open files.
Limit the number of open file descriptors per process to \f[I]n\f[R].
.TP
\f[B]\-p\f[R] \f[I]n\f[R]
limit processes.
Limit the number of processes per uid to \f[I]n\f[R].
.TP
\f[B]\-f\f[R] \f[I]bytes\f[R]
limit output size.
Limit the output file size to \f[I]bytes\f[R] bytes.
.TP
\f[B]\-c\f[R] \f[I]bytes\f[R]
limit core size.
Limit the core file size to \f[I]bytes\f[R] bytes.
.TP
\f[B]\-t\f[R] \f[I]seconds\f[R]
limit CPU time.
Limit CPU time to \f[I]seconds\f[R] seconds, delivering a SIGXCPU
thereafter.
.TP
\f[B]\-v\f[R]
verbose.
Print verbose messages to standard error.
This includes warnings about limits unsupported by the system.
.TP
\f[B]\-V\f[R]
version string.
Print a version string to standard error.
.TP
\f[B]\-P\f[R]
pgrphack.
Run \f[I]prog\f[R] in a new process group.
.TP
\f[B]\-F\f[R]
fghack or follow.
Run \f[I]prog\f[R] in new pid and mount namespaces as second process
(pid 2), and wait for \f[I]prog\f[R] and all processes started by it to
exit, no matter in which order.
Then exit with the return code of the last process exited.
The /proc filesystem is unmounted in \f[I]prog\f[R]\[aq]s namespace, and
while waiting \f[B]chpst\f[R] relays most signals it receives to all
processes in \f[I]prog\f[R]\[aq]s namespace.
On QUIT signal \f[B]chpst\f[R] sends the KILL signal to pid 1 in the
namespace to have it, \f[I]prog\f[R], and all its children exit
immediately.
Only works on systems that provide \f[B]unshare\f[R](2).
.TP
\f[B]\-I\f[R]
init.
Run \f[I]prog\f[R] in new pid and mount namespaces as process no 1 (pid
1), and wait for it to exit.
\f[I]prog\f[R] must be an init program.
The /proc filesystem is unmounted in \f[I]prog\f[R]\[aq]s namespace, and
while waiting \f[B]chpst\f[R] relays most signals it receives to
\f[I]prog\f[R].
On QUIT signal \f[B]chpst\f[R] sends the KILL signal to \f[I]prog\f[R]
to have it and all its children exit immediately.
Only works on systems that provide \f[B]unshare\f[R](2).
.TP
\f[B]\-0\f[R]
Close standard input before starting \f[I]prog\f[R].
.TP
\f[B]\-1\f[R]
Close standard output before starting \f[I]prog\f[R].
.TP
\f[B]\-2\f[R]
Close standard error before starting \f[I]prog\f[R].
.TP
\f[B]\-N\f[R]
/dev/null.
Instead of closing standard streams with \-012, re\-open the descriptors
to /dev/null.
.SH EXIT CODES
\f[B]chpst\f[R] exits 100 when called with wrong options.
It prints an error message and exits 111 if it has trouble changing the
process state.
Otherwise its exit code is the same as that of \f[I]prog\f[R].
.SH EMULATION
If \f[B]chpst\f[R] is called as \f[B]envdir\f[R], \f[B]envuidgid\f[R],
\f[B]pgrphack\f[R], \f[B]setlock\f[R], \f[B]setuidgid\f[R], or
\f[B]softlimit\f[R], it emulates the functionality of these programs
from the daemontools package respectively.
.SH SEE ALSO
sv(8), runsv(8), setsid(2), runit(8), runit\-init(8), runsvdir(8),
runsvchdir(8)
.PP
https://smarden.org/runit/
.PD 0
.P
.PD
https://cr.yp.to/daemontools.html
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * chpst.c, chpst.check, man/chpst.8: add new option -A: alarm(2) support
    (github/pilcrow).
  * sv.c: improve warning if ./log exists but not ./log/supervise/ok (thx
    github/Swiddis).
  * chpst.c, man/chpst.8: add new option -N: with -012 reopen to /dev/null;
    -2: dup() and coe() stderr, restore on failed exec (github/pilcrow).
  * man/runsvdir.8: typo.
  * doc/*.html: add color-scheme.
  * runsvdir.c: use inotify(7) (if available) to watch services directory;
    remove overcautious delay; cleanup removed services more quickly.
  * Makefile, TARGETS: add test for inotify_init() system call.
  * hasinotify.h1, hasinotify.h2, tryinotify.c: new.
  * runsvdir.c: add selfpipe; buffer and optimize readproctitle log; close
    unused filedescriptor.
  * chpst.c: -F, -I: move signal handler functions to another ifdef; do not
    relay SIGCHLD also in main pid.
  * sv.c: status: if log/supervise/ can't be accessed, don't print confusing
    additional line after warning message (introduced with 5fe1bc7).
  * etc/debian: rename to etc/sysv.
  * etc/freebsd: rename to etc/bsd.
  * etc/openbsd, etc/freebsd/getty-ttyv4, etc/debian/getty-tty5: remove;
    obsolete.
  * doc/replaceinit.html: revise sections "Replacing init on *BSD" and
    "Replacing sysvinit on GNU/Linux".
  * doc/install.html: how to install as non-root user; optionally use cc on
    *BSD.
  * man/chpst.8: document new -F and -I options.
  * chpst.c: -F option implies -I (thx github/andy-bower).
  * chpst.c: add new option -F: fghack using unshare(); chpst.c: add new
    option -I: run another init using unshare().
  * Makefile, TARGETS: add test for unshare() system call (_GNU_SOURCE).
  * hasunshare.h1, hasunshare.h2, tryunshare.c: new.
  * conf-cc: trim.
  * doc/install.html, doc/replaceinit.html, doc/runlevels.html,
    doc/useinit.html: strip '^# ' from code blocks for multiline copy&paste.
  * man/sv.8, doc/dependencies.html: rename check script to program.
  * doc/runscripts.html: clarify outdated status of the collection.
  * doc/*.html, man/*.8: replace; converted from markdown source.
  * svlogd.c, man/svlogd.8: -LL: print warning on data loss.
  * svlogd.c, man/svlogd.8: add option -L, lossy (thx github/teslamotors).
  * package/upgrade: remove utmpset program on upgrade.
  * doc/replaceinit.html: remove getty finish scripts that use utmpset;
    etc/debian/getty-tty5/finish, etc/freebsd/getty-ttyv4/finish,
    etc/openbsd/getty-ttyC4/finish: remove, obsolete.
  * doc/index.html, man/runit-init.8, man/runit.8, man/runsv.8,
    man/runsvdir.8: remove reference to utmpset program.
  * utmpset.c, utmpset.check, utmpset.dist, man/utmpset.8, tryuwtmp.c,
    tryuwtmpx.c, uw_tmp.h?: remove, obsolete.
  * Makefile, package/commands, src/Makefile, TARGETS: remove utmpset and
    utmpx sysdep.

runit 2.3.1
Sat, 07 Feb 2026 11:11:01 +0100
  * doc/dependencies.html: wording; cruft.
  * runsv.c, man/runsv.8: run control/? with service process id as argument.
  * sv.c: always write to control pipe, do not look at the status.
  * svlogd.c: fix uninitialized pointer read with -v and log directory
    mode r--.
  * Makefile, TARGETS, byte.h: add byte_zero to byte.a.
  * byte_zero.c: new.
  * runsv.c: omit stat, pid in ./supervise/ if RUNSV_STATUS_BINARYONLY is
    defined.
  * chpst.dist, sv.dist, svlogd.dist: adapt.
  * buffer_put.c: fix compilation warning: -Wempty-body (github/clan).
  * buffer_put.c, byte.h, byte_copy.c, byte_diff.c: fix compilation
    warnings: -Wdiscarded-qualifiers (github/clan).
  * sv.c: fix compilation warning: -Wdangling-else (github/clan).
  * byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c, byte_rchr.c, chpst.c,
    str_chr.c, str_len.c, sv.c, svlogd.c: fix compilation warnings:
    -Wmisleading-indentation (github/clan).
  * TARGETS: remove prot.o (github/clan).
  * doc/benefits.html: spelling.
  * doc/faq.html: cleanup, spelling.
  * doc/usedietlibc.html: remove.

src/chpst.c
#include "hasunshare.h"
#ifdef HASUNSHARE
#define _GNU_SOURCE
#include <sched.h>
#include <sys/mount.h>
#include <sys/wait.h>
#include <signal.h>
#include "sig.h"
#include "wait.h"
#endif
#include <sys/types.h>
#include <time.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <unistd.h>
#include <grp.h>
#include <fcntl.h>
#include "sgetopt.h"
#include "error.h"
#include "strerr.h"
#include "str.h"
#include "uidgid.h"
#include "strerr.h"
#include "scan.h"
#include "fmt.h"
#include "lock.h"
#include "pathexec.h"
#include "stralloc.h"
#include "byte.h"
#include "open.h"
#include "openreadclose.h"
#include "direntry.h"
#include "coe.h"
#include "fd.h"

#define USAGE_MAIN " [-vVPFI012N] [-u user[:group]] [-U user[:group]] [-b argv0] [-e dir] [-/ root] [-C pwd] [-n nice] [-l|-L lock] [-A n] [-m n] [-d n] [-o n] [-p n] [-f n] [-c n] [-t n] prog"
#define FATAL "chpst: fatal: "
#define WARNING "chpst: warning: "

const char *progname;
static stralloc sa;
char bufnum[FMT_ULONG];
#ifdef HASUNSHARE
pid_t pid;

void sig_handler(int sig) { kill(pid, sig); }
void sig_handler_pid1(int sig) { kill(-1, sig); }
void sig_quit_handler(int unused) { kill(pid, SIGKILL); }
#endif
void fatal(const char *m) { strerr_die3sys(111, FATAL, m, ": "); }
void fatal2(const char *m0, const char *m1) {
  strerr_die5sys(111, FATAL, m0, ": ", m1, ": ");
}
void fatalx(const char *m0, const char *m1) {
  strerr_die4x(111, FATAL, m0, ": ", m1);
}
void warn(const char *m) { strerr_warn2(WARNING, m, 0); }
void die_nomem() { strerr_die2x(111, FATAL, "out of memory."); }
void usage() { strerr_die4x(100, "usage: ", progname, USAGE_MAIN, "\n"); }

char *set_user =0;
char *env_user =0;
char *argv0 =0;
const char *env_dir =0;
unsigned int verbose =0;
unsigned int pgrp =0;
unsigned int devnull =0;
unsigned int nostdin =0;
unsigned int nostdout =0;
unsigned int nostderr =0;
int savederr =-1;
unsigned int newpids =0;
long limitd =-2;
long limits =-2;
long limitl =-2;
long limita =-2;
long limito =-2;
long limitp =-2;
long limitf =-2;
long limitc =-2;
long limitr =-2;
long limitt =-2;
long nicelvl =0;
const char *lock =0;
const char *root =0;
const char *pwd =0;
long alrmsecs =-1;
unsigned int lockdelay;

void suidgid(char *user, unsigned int ext) {
  struct uidgid ugid;

  if (ext) {
    if (! uidgids_get(&ugid, user)) {
      if (*user == ':') fatalx("invalid uid/gids", user +1);
      if (errno) fatal("unable to get password/group file entry");
      fatalx("unknown user/group", user);
    }
  }
  else
    if (! uidgid_get(&ugid, user)) {
      if (errno) fatal("unable to get password file entry");
      fatalx("unknown account", user);
    }
  if (setgroups(ugid.gids, ugid.gid) == -1) fatal("unable to setgroups");
  if (setgid(*ugid.gid) == -1) fatal("unable to setgid");
  if (setuid(ugid.uid) == -1) fatal("unable to setuid");
}

void euidgid(char *user, unsigned int ext) {
  struct uidgid ugid;

  if (ext) {
    if (! uidgids_get(&ugid, user)) {
      if (*user == ':') fatalx("invalid uid/gids", user +1);
      if (errno) fatal("unable to get password/group file entry");
      fatalx("unknown user/group", user);
    }
  }
  else
    if (! uidgid_get(&ugid, user)) {
      if (errno) fatal("unable to get password file entry");
      fatalx("unknown account", user);
    }
  bufnum[fmt_ulong(bufnum, *ugid.gid)] =0;
  if (! pathexec_env("GID", bufnum)) die_nomem();
  bufnum[fmt_ulong(bufnum, ugid.uid)] =0;
  if (! pathexec_env("UID", bufnum)) die_nomem();
}

void edir(const char *dirname) {
  int wdir;
  DIR *dir;
  direntry *d;
  int i;

  if ((wdir =open_read(".")) == -1)
    fatal("unable to open current working directory");
  if (chdir(dirname)) fatal2("unable to switch to directory", dirname);
  if (! (dir =opendir("."))) fatal2("unable to open directory", dirname);
  for (;;) {
    errno =0;
    d =readdir(dir);
    if (! d) {
      if (errno) fatal2("unable to read directory", dirname);
      break;
    }
    if (d->d_name[0] == '.') continue;
    if (openreadclose(d->d_name, &sa, 256) == -1) {
      if ((errno == error_isdir) && env_dir) {
        if (verbose)
          strerr_warn6(WARNING, "unable to read ", dirname, "/",
                       d->d_name, ": ", &strerr_sys);
        continue;
      }
      else
        strerr_die6sys(111, FATAL, "unable to read ", dirname, "/",
                             d->d_name, ": ");
    }
    if (sa.len) {
      sa.len =byte_chr(sa.s, sa.len, '\n');
      while (sa.len && (sa.s[sa.len -1] == ' ' || sa.s[sa.len -1] == '\t'))
        --sa.len;
      for (i =0; i < sa.len; ++i) if (! sa.s[i]) sa.s[i] ='\n';
      if (! stralloc_0(&sa)) die_nomem();
      if (! pathexec_env(d->d_name, sa.s)) die_nomem();
    }
    else
      if (! pathexec_env(d->d_name, 0)) die_nomem();
  }
  closedir(dir);
  if (fchdir(wdir) == -1) fatal("unable to switch to starting directory");
  close(wdir);
}

void slock_die(const char *m, const char *f, unsigned int x) {
  if (! x) fatal2(m, f);
  _exit(0);
}
void slock(const char *f, unsigned int d, unsigned int x) {
  int fd;

  if ((fd =open_append(f)) == -1) slock_die("unable to open lock", f, x);
  if (d) {
    if (lock_ex(fd) == -1) slock_die("unable to lock", f, x);
    return;
  }
  if (lock_exnb(fd) == -1) slock_die("unable to lock", f, x);
}

void limit(int what, long l) {
  struct rlimit r;

  if (getrlimit(what, &r) == -1) fatal("unable to getrlimit()");
  if ((l < 0) || (l > r.rlim_max))
    r.rlim_cur =r.rlim_max;
  else
    r.rlim_cur =l;
  if (setrlimit(what, &r) == -1) fatal("unable to setrlimit()");
}
void slimit() {
  if (limitd >= -1) {
#ifdef RLIMIT_DATA
    limit(RLIMIT_DATA, limitd);
#else
    if (verbose) warn("system does not support RLIMIT_DATA");
#endif
  }
  if (limits >= -1) {
#ifdef RLIMIT_STACK
    limit(RLIMIT_STACK, limits);
#else
    if (verbose) warn("system does not support RLIMIT_STACK");
#endif
  }
  if (limitl >= -1) {
#ifdef RLIMIT_MEMLOCK
    limit(RLIMIT_MEMLOCK, limitl);
#else
    if (verbose) warn("system does not support RLIMIT_MEMLOCK");
#endif
  }
  if (limita >= -1) {
#ifdef RLIMIT_VMEM
    limit(RLIMIT_VMEM, limita);
#else
#ifdef RLIMIT_AS
    limit(RLIMIT_AS, limita);
#else
    if (verbose)
      warn("system does neither support RLIMIT_VMEM nor RLIMIT_AS");
#endif
#endif
  }
  if (limito >= -1) {
#ifdef RLIMIT_NOFILE
    limit(RLIMIT_NOFILE, limito);
#else
#ifdef RLIMIT_OFILE
    limit(RLIMIT_OFILE, limito);
#else
    if (verbose)
      warn("system does neither support RLIMIT_NOFILE nor RLIMIT_OFILE");
#endif
#endif
  }
  if (limitp >= -1) {
#ifdef RLIMIT_NPROC
    limit(RLIMIT_NPROC, limitp);
#else
    if (verbose) warn("system does not support RLIMIT_NPROC");
#endif
  }
  if (limitf >= -1) {
#ifdef RLIMIT_FSIZE
    limit(RLIMIT_FSIZE, limitf);
#else
    if (verbose) warn("system does not support RLIMIT_FSIZE");
#endif
  }
  if (limitc >= -1) {
#ifdef RLIMIT_CORE
    limit(RLIMIT_CORE, limitc);
#else
    if (verbose) warn("system does not support RLIMIT_CORE");
#endif
  }
  if (limitr >= -1) {
#ifdef RLIMIT_RSS
    limit(RLIMIT_RSS, limitr);
#else
    if (verbose) warn("system does not support RLIMIT_RSS");
#endif
  }
  if (limitt >= -1) {
#ifdef RLIMIT_CPU
    limit(RLIMIT_CPU, limitt);
#else
    if (verbose) warn("system does not support RLIMIT_CPU");
#endif
  }
}

void newpid1() {
#ifdef HASUNSHARE
  int i, w =0;

  for (i =0; i < 32; ++i) sig_catch(i, sig_handler);
#ifdef SIGRTMIN
  for (i =SIGRTMIN; i <= SIGRTMAX; ++i) sig_catch(i, sig_handler);
#endif
  sig_catch(SIGQUIT, sig_quit_handler);
  sig_uncatch(sig_child);
  if (unshare(CLONE_NEWPID | CLONE_NEWNS) == -1)
    fatal("unable to set namespaces");
  if ((pid =fork()) == -1) fatal("unable to fork new pid 1");
  if (pid) { /* parent, signal relay */
    if (wait_pid(&i, pid) == -1) fatal("unable to wait for new pid 1");
    if (verbose) warn("pid1: done");
    _exit(wait_crashed(i) ? 128 + WTERMSIG(i) : wait_exitcode(i));
  }
  /* new pid 1 */
  if (verbose) warn("pid1: start");
  if (mount(NULL, "/", NULL, MS_PRIVATE|MS_REC, NULL) == -1)
    if (!root) fatal("pid1: unable to set root mount propagation");
  if (umount2("/proc", MNT_DETACH) == -1)
    if (errno != EINVAL) fatal("pid1: unable to umount /proc");
  if (newpids > 1) {
    for (i =0; i < 32; ++i) sig_catch(i, sig_handler_pid1);
#ifdef SIGRTMIN
    for (i =SIGRTMIN; i <= SIGRTMAX; ++i) sig_catch(i, sig_handler_pid1);
#endif
    sig_uncatch(sig_child);
    if ((pid =fork()) == -1) fatal("pid1: unable to fork pid 2");
    if (pid) /* parent, zombies, amialone */
      for (;;) {
        pid_t p;
        if ((p =wait_pid(&i, -1)) == -1) {
          if (errno != ECHILD)
            strerr_warn2(WARNING, "pid1: wait_pid(): ", &strerr_sys);
          _exit(wait_crashed(w) ? 128 + WTERMSIG(w) : wait_exitcode(w));
        } else {
          w =i;
          if (verbose) {
            bufnum[fmt_ulong(bufnum, p)] =0;
            strerr_warn4(WARNING, "pid1: pid ", bufnum, ": exit", 0);
          }
        }
      }
    /* pid 2 */
  }
  for (i =0; i < 32; ++i) sig_uncatch(i);
#ifdef SIGRTMIN
  for (i =SIGRTMIN; i <= SIGRTMAX; ++i) sig_uncatch(i);
#endif
#else
  fatalx("unable to set namespaces", "system does not provide unshare()");
#endif
}

int dismiss_fd(int fd) {
  int dn;

  if (! devnull) return close(fd);

  if ((dn =open("/dev/null", O_RDWR)) == -1) return -1;
  return fd_move(fd, dn);
}

/* argv[0] */
void setuidgid(int, char *const *);
void envuidgid(int, char *const *);
void envdir(int, char *const *);
void pgrphack(int, char *const *);
void setlock(int, char *const *);
void softlimit(int, char *const *);

int main(int argc, char **argv) {
  int opt;
  int i;
  unsigned long ul;

  progname =argv[0];
  for (i =str_len(progname); i; --i)
    if (progname[i -1] == '/') {
      progname +=i;
      break;
    }
  if (progname[0] == 'd') ++progname;

  /* argv[0] */
  if (str_equal(progname, "setuidgid")) setuidgid(argc, argv);
  if (str_equal(progname, "envuidgid")) envuidgid(argc, argv);
  if (str_equal(progname, "envdir")) envdir(argc, argv);
  if (str_equal(progname, "pgrphack")) pgrphack(argc, argv);
  if (str_equal(progname, "setlock")) setlock(argc, argv);
  if (str_equal(progname, "softlimit")) softlimit(argc, argv);

  while ((opt =getopt(argc, argv, "u:U:b:e:m:d:o:p:f:c:r:t:/:C:n:l:L:A:vP012NFIV"))
         != opteof)
    switch(opt) {
    case 'u': set_user =(char*)optarg; break;
    case 'U': env_user =(char*)optarg; break;
    case 'b': argv0 =(char*)optarg; break;
    case 'e': env_dir =optarg; break;
    case 'm':
      if (optarg[scan_ulong(optarg, &ul)]) usage();
      limits =limitl =limita =limitd =ul;
      break;
    case 'd': if (optarg[scan_ulong(optarg, &ul)]) usage(); limitd =ul; break;
    case 'o': if (optarg[scan_ulong(optarg, &ul)]) usage(); limito =ul; break;
    case 'p': if (optarg[scan_ulong(optarg, &ul)]) usage(); limitp =ul; break;
    case 'f': if (optarg[scan_ulong(optarg, &ul)]) usage(); limitf =ul; break;
    case 'c': if (optarg[scan_ulong(optarg, &ul)]) usage(); limitc =ul; break;
    case 'r': if (optarg[scan_ulong(optarg, &ul)]) usage(); limitr =ul; break;
    case 't': if (optarg[scan_ulong(optarg, &ul)]) usage(); limitt =ul; break;
    case 'A': if (optarg[scan_ulong(optarg, &ul)]) usage(); alrmsecs =(long)ul; break;
    case '/': root =optarg; break;
    case 'C': pwd =optarg; break;
    case 'n':
      switch (*optarg) {
        case '-':
          ++optarg;
          if (optarg[scan_ulong(optarg, &ul)]) usage();
          nicelvl =-ul;
          break;
        case '+': ++optarg;
        default:
          if (optarg[scan_ulong(optarg, &ul)]) usage();
          nicelvl =ul;
          break;
      }
      break;
    case 'l': if (lock) usage(); lock =optarg; lockdelay =1; break;
    case 'L': if (lock) usage(); lock =optarg; lockdelay =0; break;
    case 'v': verbose =1; break;
    case 'N': devnull =1; break;
    case 'P': pgrp =1; break;
    case '0': nostdin =1; break;
    case '1': nostdout =1; break;
    case '2': nostderr =1; break;
    case 'F': newpids =2; break;
    case 'I': if (!newpids) newpids =1; break;
    case 'V': strerr_warn1("$Id$", 0);
    case '?': usage();
    }
  argv +=optind;
  if (! argv || ! *argv) usage();

  if (pgrp) setsid();
  if (env_dir) edir(env_dir);
  if (root) {
    if (chdir(root) == -1) fatal2("unable to change directory", root);
    if (chroot(".") == -1) fatal("unable to change root directory");
  }
  if (pwd) {
    if (chdir(pwd) == -1) fatal2("unable to change directory", pwd);
  }
  if (nicelvl) {
    errno =0;
    if (nice(nicelvl) == -1) if (errno) fatal("unable to set nice level");
  }
  if (alrmsecs >= 0) alarm((unsigned int)alrmsecs);
  if (newpids) newpid1();
  if (env_user) euidgid(env_user, 1);
  if (set_user) suidgid(set_user, 1);
  if (lock) slock(lock, lockdelay, 0);
  if (nostderr) {
    if ((savederr = dup(2)) == -1) fatal("unable to save stderr");
    coe(savederr);
    if (dismiss_fd(2) == -1) fatal("unable to close stderr");
  }
  if (nostdout) if (dismiss_fd(1) == -1) fatal("unable to close stdout");
  if (nostdin) if (dismiss_fd(0) == -1) fatal("unable to close stdin");
  slimit();

  progname =*argv;
  if (argv0) *argv =argv0;
  pathexec_env_run(progname, argv);
  if (savederr >= 0) {
    int e = errno;
    fd_move(2, savederr);
    errno = e;
  }
  fatal2("unable to run", *argv);
  return(0);
}

/* argv[0] */
#define USAGE_SETUIDGID " account child"
#define USAGE_ENVUIDGID " account child"
#define USAGE_ENVDIR " dir child"
#define USAGE_PGRPHACK " child"
#define USAGE_SETLOCK " [ -nNxX ] file program [ arg ... ]"
#define USAGE_SOFTLIMIT " [-a allbytes] [-c corebytes] [-d databytes] [-f filebytes] [-l lockbytes] [-m membytes] [-o openfiles] [-p processes] [-r residentbytes] [-s stackbytes] [-t cpusecs] child"

void setuidgid_usage() {
  strerr_die4x(100, "usage: ", progname, USAGE_SETUIDGID, "\n");
}
void setuidgid(int argc, char *const *argv) {
  const char *account;

  if (! (account =*++argv)) setuidgid_usage();
  if (! *++argv) setuidgid_usage();
  suidgid((char*)account, 0);
  pathexec(argv);
  fatal2("unable to run", *argv);
}

void envuidgid_usage() {
  strerr_die4x(100, "usage: ", progname, USAGE_ENVUIDGID, "\n");
}
void envuidgid(int argc, char *const *argv) {
  const char *account;

  if (! (account =*++argv)) envuidgid_usage();
  if (! *++argv) envuidgid_usage();
  euidgid((char*)account, 0);
  pathexec(argv);
  fatal2("unable to run", *argv);
}

void envdir_usage() {
  strerr_die4x(100, "usage: ", progname, USAGE_ENVDIR, "\n");
}
void envdir(int argc, char *const *argv) {
  const char *dir;

  if (! (dir =*++argv)) envdir_usage();
  if (! *++argv) envdir_usage();
  edir(dir);
  pathexec(argv);
  fatal2("unable to run", *argv);
}

void pgrphack_usage() {
  strerr_die4x(100, "usage: ", progname, USAGE_PGRPHACK, "\n");
}
void pgrphack(int argc, char *const *argv) {
  if (! *++argv) pgrphack_usage();
  setsid();
  pathexec(argv);
  fatal2("unable to run", *argv);
}

void setlock_usage() {
  strerr_die4x(100, "usage: ", progname, USAGE_SETLOCK, "\n");
}
void setlock(int argc, char *const *argv) {
  int opt;
  unsigned int delay =0;
  unsigned int x =0;
  const char *fn;

  while ((opt =getopt(argc, argv, "nNxX")) != opteof)
    switch(opt) {
      case 'n': delay =1; break;
      case 'N': delay =0; break;
      case 'x': x =1; break;
      case 'X': x =0; break;
      default: setlock_usage();
    }
  argv +=optind;
  if (! (fn =*argv)) setlock_usage();
  if (! *++argv) setlock_usage();

  slock(fn, delay, x);
  pathexec(argv);
  if (! x) fatal2("unable to run", *argv);
  _exit(0);
}

void softlimit_usage() {
  strerr_die4x(100, "usage: ", progname, USAGE_SOFTLIMIT, "\n");
}
void getlarg(long *l) {
  unsigned long ul;

  if (str_equal(optarg, "=")) { *l =-1; return; }
  if (optarg[scan_ulong(optarg, &ul)]) usage();
  *l =ul;
}
void softlimit(int argc, char *const *argv) {
  int opt;
  
  while ((opt =getopt(argc,argv,"a:c:d:f:l:m:o:p:r:s:t:")) != opteof)
    switch(opt) {
    case '?': softlimit_usage();
    case 'a': getlarg(&limita); break;
    case 'c': getlarg(&limitc); break;
    case 'd': getlarg(&limitd); break;
    case 'f': getlarg(&limitf); break;
    case 'l': getlarg(&limitl); break;
    case 'm': getlarg(&limitd); limits =limitl =limita =limitd; break;
    case 'o': getlarg(&limito); break;
    case 'p': getlarg(&limitp); break;
    case 'r': getlarg(&limitr); break;
    case 's': getlarg(&limits); break;
    case 't': getlarg(&limitt); break;
    }
  argv +=optind;
  if (!*argv) softlimit_usage();
  slimit();
  pathexec(argv);
  fatal2("unable to run", *argv);
}
